### DISCUSSION TITLE
Using Neuromancer for Moving Horizon Estimation

### DISCUSSION THREAD

**Discussion by hadlaczkyb** [pnnl/neuromancer#87] [Q&A] [answered]
Hi!

I would like to ask for some help in the following matter.
I am trying to use Neuromancer for learning-based/differenciable moving horizon estimation in a similar fashion as it is used for the DMPC.
However, though the optimization problem of the MHE is quite similar to the MPC's in its formalization, there are some differences: here the current value of the optimization variable - the estimated state vector (`x_k`) - is compared to its value at a different time step (see equations - `w_k`).

![mhe](https://github.com/pnnl/neuromancer/assets/72016483/bb9682a0-b348-4d4b-9501-75380448fe3e)

And this part is what I could not really implement/work out how to implement in Neuromancer, as I did not find such an example where an optimization variable is used in such a way. (_Or maybe I have missed it..._)
So my question is: is it possible to implement this using Neuromancer 'out-of-the-box' (_so without extending the functionality of any of the existing classes_)? And if it is possible, could you give me some help/suggestions how to do it?

Thanks in advance!

**Comment by drgona** [comment, marked as answer]
Hi @hadlaczkyb thank you for your question.
Yes it is possible to implement MHE using neuromancer. You can adapt one of our differentiable predictive control examples for this purpose.
In your case, $y_k$ will be a sampled trajectory comming from the dataset. Then you need to implement the state space model equations and instantiate neural network that will map the problem parameters to the decition variables.
https://github.com/pnnl/neuromancer/tree/master/examples/control
